dbms/src/Flash/Mpp/ExchangeReceiverCommon.h

```diff
@@ -0,0 +1,429 @@
+{
+    GET_METRIC(tiflash_exchange_queueing_data_bytes, type_receive).Decrement(data_size_in_queue.load());
+}
+} // namespace ExchangeReceiverMetric
+
+constexpr Int32 batch_packet_count = 16;
+
+struct ReceivedMessage
+{
+    size_t source_index;
+    String req_info;
+    // shared_ptr<const MPPDataPacket> is copied to make sure error_ptr, resp_ptr and chunks are valid.
+    const std::shared_ptr<DB::TrackedMppDataPacket> packet;
+    const mpp::Error * error_ptr;
+    const String * resp_ptr;
+    std::vector<const String *> chunks;
+
+    // Constructor that move chunks.
+    ReceivedMessage(size_t source_index_,
+                    const String & req_info_,
+                    const std::shared_ptr<DB::TrackedMppDataPacket> & packet_,
+                    const mpp::Error * error_ptr_,
+                    const String * resp_ptr_,
+                    std::vector<const String *> && chunks_)
+        : source_index(source_index_)
+        , req_info(req_info_)
+        , packet(packet_)
+        , error_ptr(error_ptr_)
+        , resp_ptr(resp_ptr_)
+        , chunks(chunks_)
+    {}
+
+    void switchMemTracker()
+    {
+        packet->switchMemTracker(current_memory_tracker);
+    }
+};
+
+struct ExchangeReceiverResult
+{
+    std::shared_ptr<tipb::SelectResponse> resp;
+    size_t call_index;
+    String req_info;
+    bool meet_error;
+    String error_msg;
+    bool eof;
+    DecodeDetail decode_detail;
+
+    ExchangeReceiverResult()
+        : ExchangeReceiverResult(nullptr, 0)
+    {}
+
+    static ExchangeReceiverResult newOk(std::shared_ptr<tipb::SelectResponse> resp_, size_t call_index_, const String & req_info_)
+    {
+        return {resp_, call_index_, req_info_, /*meet_error*/ false, /*error_msg*/ "", /*eof*/ false};
+    }
+
+    static ExchangeReceiverResult newEOF(const String & req_info_)
+    {
+        return {/*resp*/ nullptr, 0, req_info_, /*meet_error*/ false, /*error_msg*/ "", /*eof*/ true};
+    }
+
+    static ExchangeReceiverResult newError(size_t call_index, const String & req_info, const String & error_msg)
+    {
+        return {/*resp*/ nullptr, call_index, req_info, /*meet_error*/ true, error_msg, /*eof*/ false};
+    }
+
+private:
+    ExchangeReceiverResult(
+        std::shared_ptr<tipb::SelectResponse> resp_,
+        size_t call_index_,
+        const String & req_info_ = "",
+        bool meet_error_ = false,
+        const String & error_msg_ = "",
+        bool eof_ = false)
+        : resp(resp_)
+        , call_index(call_index_)
+        , req_info(req_info_)
+        , meet_error(meet_error_)
+        , error_msg(error_msg_)
+        , eof(eof_)
+    {}
+};
+
+enum class ExchangeReceiverState
+{
+    NORMAL,
+    ERROR,
+    CANCELED,
+    CLOSED,
+};
+
+using MsgChannelPtr = std::shared_ptr<MPMCQueue<std::shared_ptr<ReceivedMessage>>>;
+
+template <bool is_sync>
+inline void injectFailPointReceiverPushFail(bool & push_succeed)
+{
+    if constexpr (is_sync)
```

Comment:
It looks like we can add an enum and add a failpoint to LocalTunnel and add the relevant tests.
```
switch (mode)
{
case Local:
    fiu_do_on(FailPoints::random_receiver_local_msg_push_failure_failpoint, push_succeed = false;);
    break;
case Sync:
    fiu_do_on(FailPoints::random_receiver_sync_msg_push_failure_failpoint, push_succeed = false;);
    break
case Async:
    fiu_do_on(FailPoints::random_receiver_async_msg_push_failure_failpoint, push_succeed = false;);
    break;
}
```

Reply:
And then we can add the `random_receiver_local_msg_push_failure_failpoint` in random fail point.